### DISCUSSION TITLE
On error event is only fired after connection is restored

### DISCUSSION THREAD

**Discussion by cgvalayev** [enisdenjo/graphql-ws#431] [Q&A] [answered]
Steps to reproduce:
1. Chrome dev tools "Network conditions -> Offline" - we get nothing from the graphql client [on.error, on.closed](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#on)
2. "Network conditions -> No throttling" - on.error fires

Also, I am having a hard time understanding when the client is trying to retry and how many attempts left.

Use cases: 
1. **I would like to be able to display connectivity warning in my app, should I be using something else?**
2. What is the best way to handle reconnection after all attempts were exceeded, for example user left computer for 2 days and came back? - answered here https://github.com/enisdenjo/graphql-ws/discussions/121

**Comment by enisdenjo** [comment, marked as answer]
`on.error` and `on.closed` are hooked up directly to the browsers WebSocket implementation, see here:

https://github.com/enisdenjo/graphql-ws/blob/48e7bd931ebe58e4689d7cdfcfa3e413b1a23062/src/client.ts#L665-L666

If playing with network throttling does not yield the expected results, I reckon it's due to some browser thing. You can confirm this by using the WebSocket directly.

> Also, I am having a hard time understanding when the client is trying to retry and how many attempts left.

The [retryAttempts](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#retryattempts) client option controls the amount of retries and has the details answering your question. Furthermore, also read about the [shouldRetry](https://github.com/enisdenjo/graphql-ws/blob/master/docs/interfaces/client.ClientOptions.md#shouldretry) client option.

> I would like to be able to display connectivity warning in my app, should I be using something else?

1. [on.connecting](https://github.com/enisdenjo/graphql-ws/blob/master/docs/modules/client.md#eventconnecting) indicates that the connection process to the server has started
1. [on.connected](https://github.com/enisdenjo/graphql-ws/blob/master/docs/modules/client.md#eventconnected) indicates that the client has opened a WebSocket connection and that the server acknowledged
1. [on.closed](https://github.com/enisdenjo/graphql-ws/blob/master/docs/modules/client.md#eventclosed) indicates that the WebSocket connection has closed, for whatever reason
